tracing-journald/src/socket.rs

```diff
@@ -0,0 +1,59 @@
+//! socket helpers.
+
+use std::io::{Error, Result};
+use std::mem::size_of;
+use std::os::unix::net::UnixDatagram;
+use std::os::unix::prelude::{AsRawFd, RawFd};
+use std::ptr;
+
+use libc::*;
+
+const CMSG_BUFSIZE: usize = 64;
+
+#[repr(C)]
+union AlignedBuffer<T: Copy + Clone> {
+    buffer: T,
+    align: cmsghdr,
+}
+
+fn cmsg_space(s: u32) -> usize {
+    unsafe { CMSG_SPACE(s) as usize }
+}
+
+pub fn send_one_fd(socket: &UnixDatagram, fd: RawFd) -> Result<usize> {
+    assert!(
+        cmsg_space(size_of::<RawFd>() as u32) <= CMSG_BUFSIZE,
+        "cmsghdr buffer too small (< {}) to hold a single fd",
+        cmsg_space(size_of::<RawFd>() as u32)
+    );
+
+    let mut cmsg_buffer = AlignedBuffer {
+        buffer: ([0u8; CMSG_BUFSIZE]),
+    };
+    let mut msg: msghdr = unsafe { std::mem::zeroed() };
+
+    // We send no data body with this message.
+    msg.msg_iov = ptr::null_mut();
+    msg.msg_iovlen = 0;
+
+    msg.msg_control = unsafe { cmsg_buffer.buffer.as_mut_ptr() as _ };
+    msg.msg_controllen = unsafe { CMSG_SPACE(size_of::<RawFd>() as _) as _ };
+
+    let mut cmsg: &mut cmsghdr =
+        unsafe { CMSG_FIRSTHDR(&msg).as_mut() }.expect("Control message buffer exhausted");
+
+    cmsg.cmsg_level = SOL_SOCKET;
+    cmsg.cmsg_type = SCM_RIGHTS;
+    cmsg.cmsg_len = unsafe { CMSG_LEN(size_of::<RawFd>() as _) as _ };
+
+    unsafe { ptr::write(CMSG_DATA(cmsg) as *mut RawFd, fd) };
+
+    let result = unsafe { libc::sendmsg(socket.as_raw_fd(), &msg, libc::MSG_NOSIGNAL) };
```

Comment:
very minor style nit: i notice the `libc::sendmsg` is prefixed with `libc` here, but elsewhere in this file, other libc functions are called without the prefix (via the `use libc::*;` import). i have a slight preference to make this consistent (though this isn't a big deal).